File: Blind_75/best-time-to-buy-and-sell-stock.java
// 121. Best Time to Buy and Sell Stock
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

class Solution {
    public int maxProfit(int[] prices) {
        int minPrice = Integer.MAX_VALUE;
        int profit = 0;

        for (int currPrice : prices) {
            profit = Math.max(profit, currPrice - minPrice);
            minPrice = Math.min(currPrice, minPrice);
        }

        return profit;
    }
}

// init minPrice to MaxInt
// init profit to 0
// iterate through the prices
// update profit to max(profit, currPrice - minPrice)
// if currPrice < minPrice -> update the minPrice

File: best_time_to_buy_and_sell_stock.js
// 121. Best Time to Buy and Sell Stock
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let lowestPrice = Number.MAX_SAFE_INTEGER;
  let maxProfit = 0;

  for (const currentPrice of prices) {
    lowestPrice = Math.min(currentPrice, lowestPrice);
    maxProfit = Math.max(currentPrice - lowestPrice, maxProfit);
  }

  return maxProfit;
};

// init lowestPrice as largest integer
// init maxProfit as 0
// iterate through the prices
// update the lowestPrice if the price is lower than the min
// update the maxProfit if the currentPrice - lowestPrice is greater than maxProfit
